### DISCUSSION TITLE
Immich app unusably slow on Android with ~15.000 photos on the phone

### DISCUSSION THREAD

**Discussion by 4f1sh3r** [immich-app/immich#9614] [Q&A]
### The bug

Dear Immich-Team,

first - thanks for this great piece of software! It works very good so far, but there is one annoying thing, which breaks the usage for my wife.

She uses a Samsung Galaxy S20 and has ~15.000 photos on the local SD card and all of them are synced to Immich. I **_assume_** this is the reason why the Immich App on her phone is almost unusable. It's getting stuck all the time: during scrolling, during photo viewing, basically during all activities within the app. 

This breaks our use case: We create shared albums e.g. for vacations. When I share some pictures from my phone into this album, it takes several hours until it eventually shows up on her end. The other way round is not much better: Although the photos have been uploaded and are already visible on the Immich website, her app still doesn't recognize them as "uploaded" and therefore doesn't allow to put it into albums. The next thing is that she needs to delete photos in Immich in order to have them deleted in both Immich and local album. This is almost impossible as the app always hangs. Sometimes this results in an inconsistent situation where photos that have already been deleted are still displayed with an exclamation mark (see screenshots). This fixes itself after a few hours. 

All in all, the app is therefore not usable for my wife. Those problems are frustrating for her and reduces the WAF quite a bit. :-)
I know it may be an edge-case that she has so many local photos. Nevertheless, I think that the app should be able to handle a large number of local photos too.

I don't have any of those problems on my phone (iPhone 14), but I only have around 1000 pictures locally.
Please find attached the recent log file from her phone, with lots of errors in it.

We're using the latest Immich App version 1.105.0 on both phones. Immich server is on latest 1.105.1. Server and website are working fairly well and so does the app on my phone. Its just on her phone.

Please find attached the latest logfile from her app, as well as some screenshots to illustrate the issues.

Is this a well-known issue? Do you have any ideas how to improve the situation?
If you need more (or more detailed) information, please just tell me and I'll happily provide it.
Thanks in advance!

![b_WhatsApp Image 2024-05-20 at 15 03 24](https://github.com/immich-app/immich/assets/3988337/9ab1e966-b759-48c9-b939-d8f9a1a0bd0b)
![b_WhatsApp Image 2024-05-20 at 15 03 48](https://github.com/immich-app/immich/assets/3988337/73fd230a-a3f2-4f87-b20f-81048c6a08a9)
![WhatsApp Image 2024-05-20 at 14 56 58](https://github.com/immich-app/immich/assets/3988337/185da4ec-d126-489a-b4ce-9285d7f605cc)
[Immich_log_2024-05-20T14:55:30.808311.log](https://github.com/immich-app/immich/files/15378278/Immich_log_2024-05-20T14.55.30.808311.log)

### The OS that Immich Server is running on

Ubuntu 22.04

### Version of Immich Server

v1.105.1

### Version of Immich Mobile App

v1.105.0

### Platform with the issue

- [ ] Server
- [ ] Web
- [X] Mobile

### Your docker-compose.yml content

```YAML
not needed, as the server runs great
```

### Your .env content

```Shell
not needed, as the server runs great
```

### Reproduction steps

```bash
1.) Have 15.000 photos on an android phone
2.) Install Immich and sync them to Immich
3.) Immich app gets very slow
```

### Relevant log output

```shell
look at the attachment to see the whole file, here are some lines I find very suspect:

# Loading assets takes 42 seconds
2024-05-20 14:42:42.460933 | INFO     | SyncService          | Upserted 18 assets into the DB |
2024-05-20 14:42:15.846355 | INFO     | AssetNotifier        | Load assets: 41256ms |
2024-05-20 14:42:15.846283 | INFO     | AlbumService         | Syncing completed. Changes: true |


# Sometimes even longer:
2024-05-02 16:49:43.451991 | INFO     | AssetNotifier        | Load partner assets: 1054ms |
2024-05-02 16:49:42.404474 | INFO     | BackupNotifier       | [_resumeBackup] Start back up |
2024-05-02 16:16:33.969330 | INFO     | AssetNotifier        | Load partner assets: 265461ms |
2024-05-02 16:16:33.964151 | INFO     | AssetNotifier        | Load assets: 265435ms |
2024-05-02 16:16:33.963638 | INFO     | AlbumService         | Syncing completed. Changes: true |


# Sometimes flutter errors:

2024-05-18 16:17:04.868065 | SEVERE   | ImmichErrorLogger    | PlatformDispatcher - Catch all | ApiException 400: {"message":"Not found or no album.read access","error":"Bad Request","statusCode":400} |
#0      AlbumApi.getAlbumInfo (package:openapi/api/album_api.dart:313)
<asynchronous suspension>
#1      AlbumService.refreshRemoteAlbums.<anonymous closure> (package:immich_mobile/services/album.service.dart:164)
<asynchronous suspension>
#2      SyncService._addAlbumFromServer (package:immich_mobile/services/sync.service.dart:436)
<asynchronous suspension>
#3      diffSortedLists (package:immich_mobile/utils/diff.dart:21)
<asynchronous suspension>
#4      SyncService._syncRemoteAlbumsToDb (package:immich_mobile/services/sync.service.dart:306)
<asynchronous suspension>
#5      AsyncMutex.run.<anonymous closure> (package:immich_mobile/utils/async_mutex.dart:15)
<asynchronous suspension>

2024-05-18 16:17:04.867917 | SEVERE   | ImmichErrorLogger    | PlatformDispatcher - Catch all | ApiException 400: {"message":"Not found or no album.read access","error":"Bad Request","statusCode":400} |
#0      AlbumApi.getAlbumInfo (package:openapi/api/album_api.dart:313)
<asynchronous suspension>
#1      AlbumService.refreshRemoteAlbums.<anonymous closure> (package:immich_mobile/services/album.service.dart:164)
<asynchronous suspension>
#2      SyncService._addAlbumFromServer (package:immich_mobile/services/sync.service.dart:436)
<asynchronous suspension>
#3      diffSortedLists (package:immich_mobile/utils/diff.dart:21)
<asynchronous suspension>
#4      SyncService._syncRemoteAlbumsToDb (package:immich_mobile/services/sync.service.dart:306)
<asynchronous suspension>
#5      AsyncMutex.run.<anonymous closure> (package:immich_mobile/utils/async_mutex.dart:15)
<asynchronous suspension>


# and a lot more errors in the log file attached...
```

**Comment by bo0tzz** [comment]
I think this might be a duplicate of #2128. Can you try enabling "Prefer remote images" in the app settings?

**Comment by alextran1502** [comment]
15000 photos on the phone should be fine. Are you accessing over local IP or DNS? Can you try to have her phone access to the local IP?

**Comment by 4f1sh3r** [comment]
> 15000 photos on the phone should be fine. Are you accessing over local IP or DNS? Can you try to have her phone access to the local IP?

Thats not an easy task to be honest, as Immich runs behind a caddy reverse proxy on my public server. Caddy routes only request to that very hostname to Immich. If you have a reasonable suspicion, I am happy to rebuild it temporarily to test it out, but I cannot really imagine that it would fix anything tbh. I am using the same FQDN on my phone in the same WIFI network and I don't have any of those problems.


> I think this might be a duplicate of #2128. Can you try enabling "Prefer remote images" in the app settings?

Thats a good hint, I will try that tomorrow and report back. The log shows loading times of up to 265 seconds, so the workaround should definitely improve the situation. The strange thing is that it's not just the loading of the thumbnails, but that the whole app is unusable. Anyways, I'll test and report back.

Thanks to both of you for your ideas!

**Comment by 4f1sh3r** [comment]
I can confirm that enabling „prefer remote images“ improved the situation a lot! There are still situations where the app freezes for 5-10 seconds but at least it got better. Thanks for the hint. Thumbs pressed that you are able to find the root cause behind the bug. Thanks for the great work!

**Comment by klejejs** [comment]
I observed the same issue on iOS. Enabling „prefer remote images“ helped a lot as well but I still sometimes get stutters.

**Comment by matejkramny** [comment]
If you tried the same account on the iphone, i would guess it's that the sd card itself is comparably slower to onboard storage which causes the app to lag. You could try moving the photos onto the phone or get a faster sd card (or maybe you already have a decent one)?

**Comment by 4f1sh3r** [comment]
Could actually be a reason. But since the whole UI freezes, no matter where you are in the app, and the whole thing is reproducible, I'm now thinking more in the direction of your bug. https://github.com/immich-app/immich/issues/10030

Maybe it's a combination: maybe the huge DB query locks the UI. Perhaps it is not noticeable in many scenarios because there are not so many local photos. Anyway, my gut feeling is that decoupling the DB thread from the UI (if I'm saying this correctly) might help in my case too.

**Reply by matejkramny** [reply]
Some more discussion around the bug in my WIP PR implementing it: https://github.com/immich-app/immich/pull/10668

I've also tested this a bit more by downloading about 100k demo photos to my immich server and dummy user - the app seemed fine (compared to my ~80k photos).
I don't think it's the number of photos, but maybe some photos are special or it's the number of videos.. I'm at a loss though as debugging doesn't help when the app freezes / flutter crashes.

Doing the DB queries (and API calls, parsing) in another thread is the way forward as the maintainers suggested, but it's not a trivial task.

My hunch are the reactive components watching the database and perhaps doing too many refreshes or something

**Comment by smiba** [comment]
I'm dealing with a similar issue where scrolling through my photos takes seconds if not longer on Android. I don't think this was the case months ago, enabling "Prefer Remote Images" in the settings made it a lot faster again... But why is this? 

I'm sure my Samsung Flip 5 storage is faster than my cellular 

Only thing I've recently changed is my preview settings in the Administration panel of Immich, from JPEG to WEBP. Maybe these take significantly longer to decode?

**Reply by alextran1502** [reply]
The reason I can think of is the native API to get the system file on the phone is somehow very slow

**Reply by smiba** [reply]
The phone is also getting very warm scrolling through while its loading the files, its like its spending forever on decoding. Eventually after a few seconds you'll see a really pixelated thumbnail, and another few seconds later the higher resolution thumbnail appears.

It's almost as if it's having to load the full resolution of every image and generate a thumbnail for it, every single time.  

I do want to note that I dont think it's slow on individual requests, but if you scroll through your photos it quickly gets to the point where it suddenly had to generate thumbnails for over 50 photos. Especially considering I have 5 photos per row configured 

EDIT: The longer I investigate this issue, the more certain I get that the app is no longer properly storing thumbnails and is generating them for local files on the fly every single time. If I open and close the app it goes through the same rendering process every time.

**Reply by smiba** [reply]
I've made a video showing the issue here: https://youtu.be/NHyzHY1YqDw

**Reply by ericmuijs** [reply]
I had the same issue. Have more than 100K photos and videos. On my S23 lots of stutters when scrolling and phone become very hot. Better with prefer remote. Still the app is very stuttery somehow. Fresh install and probably my server is still processing some of the images. 

Love the app and the platform though, I'm truly impressed.

**Reply by matejkramny** [reply]
> The reason I can think of is the native API to get the system file on the phone is somehow very slow

@alextran1502 That can't be the case when [the app that shan't be named] doesn't have this problem. I think it might be with the plugin that handles images, it has a native counterpart in (kotlin/java, swift) and there's a Flutter bridge that it communicates with. That may be the bottleneck, either due to the fact that Flutter isolates are single-thread or it's doing some encoding/decoding in Flutter to resize the image.

Remote images might not have this problem even if it's slower to get the actual image, it might not be suffering the same issue.

**Reply by smiba** [reply]
> I've made a video showing the issue here: https://youtu.be/NHyzHY1YqDw

@alextran1502 I was wondering if you've seen this video demonstrating the issue, as I think it shows it quite well

**Reply by alextran1502** [reply]
@smiba yes, I can see it on my test Android device

**Comment by Acidiphilium** [comment]
I have the same problem and also looks like it tries to regenerate thumbnails on each immich start. Like it doesn't store them on the phone for faster rendering. It makes the app quite "slow" or let's say not very responsive compared to Google Photos. On my phone it's still usable as if I wait several seconds (around 5sec) then all my 20k photos thumbs are then appearing quickly when I scroll even if I scroll far it takes only around 1-2 secs no more (but again not as quick as Google Photos which has almost no delay, like perfectly fluid). So usable yes but not perfect.

But but the option "Prefer remote images" is the workaround. No more 5sec delay at immich start and scroll is almost like without any delay.

**Comment by Gitonthegrid** [comment]
I have the same sort of problems, but then with "people".

If I click on a person, it takes forever to load them..
The normal pictures are "fine" with loading. Not the best, but fine.

**Comment by kevincox** [comment]
I'm seeing serious jank with a library of 73k photos and 6k videos on 1.117.0 build.162

My server is on the local network (not that it should matter for UI jank)

Scrolling just about anywhere on the app is incredibly slow. Only brief moments of 60fps, regularly under 5 with some artists if about a second. 

https://youtu.be/mGZRz-91jRk

I noted that it is particularly bad on the map view. It locks up when loading and even after scrolling is super janky. It is so bad that it may be an additional issue but may be related. 

https://youtu.be/cO0OYiwPaQI

I haven't seen any freezing it crashing yet though.

**Reply by alextran1502** [reply]
@kevincox the UI jank is due to the album sync process that running on the main thread. We are  working toward a better sync mechanism which will offload most of the calculation to the server so it will reduce the jankiness sharply

**Reply by kevincox** [reply]
Great to hear. I have a bit over 200 albums if that is part of the problem. 

Is there somewhere I can subscribe to this work?

**Reply by Cohinem** [reply]
@alextran1502 That is nice. I've been having problems with very slow and laggy UI while selecting many albums with tens of thousands of images with syncing (so new pictures get uploaded automatically). The UI just becomes unuseable. So I just stopped using the android app for anything other than uploading and use the Web UI for everything else. Good to hear that you're aware and working on it!

**Reply by 4f1sh3r** [reply]
@alextran1502 Oh that's great news! Could you share the link to the WIP PR or branch once ready? I am very looking forward to it 😃 thanks again for all the amazing work!

**Reply by alextran1502** [reply]
@AleXburnA it is a series of refactoring that have been going on. I am planning to work on the album sync mechanism this coming week

**Reply by Cohinem** [reply]
Not fixed yet as of 1.19.1

**Reply by mustermaxs** [reply]
I don't know if this problem has been solved or not. But I'm still encountering this problem. The app freezes up every couple of seconds for at least 7s. I only have about 9000 photos and videos in my account.

**Reply by 4f1sh3r** [reply]
We also still experience the issue on Android phones, it seems it's not yet fixed

**Reply by cipriantudoran** [reply]
Hello! I am experiencing the slowliness issue too. 20k pictures (all on external storage, if that counts), 200 albums. Scrolling / searching albums is nearly impossible, thumb display is slow. It's almost unusable, on Android. Latest server and app. The web interface works flawlessly.

**Reply by dstapp** [reply]
Same thing here: ~25k photos, latest Immich version on Android & the server. 
Pixel 9 Pro Fold (so should not be a hardware limitation) with GrapheneOS. App is extremely unresponsive and stuttery, especially when scrolling the timeline. If that helps: It seems like that only happens when the app is fresh started after being kicked out of memory by the OS, so basically whenever you see the startup Immich logo. When it's already running in the background and I'm switching back to it after 1-2 mins in another app, everything is smooth as expected.

"Prefer images from server" does not seem to change anything in my case, but fwiw I've turned it on.

If you need any further debugging information, let me know, I'd be happy to help.

**Reply by Cohinem** [reply]
I think it's in the process of syncing when it's scanning the files. When it has been ~2mins the scanning is already complete depending on your library and device.
What I do to circumvent the problem is have 2 copies of immich app. one solely for syncing and the other for browsing. 
I don't know why they haven't fixed it yet. 
I don't experience any of these problems on Ente Photos. Maybe sneak a peek at their code and see how they approach syncing?

**Reply by alextran1502** [reply]
> I don't know why they haven't fixed it yet. 
> 

Oh we are fixing it 😅 basically a re-write of the data structure we are using in the app so it will take a bit to get done 😀

**Reply by matticas** [reply]
> > I don't know why they haven't fixed it yet.
> 
> Oh we are fixing it 😅 basically a re-write of the data structure we are using in the app so it will take a bit to get done 😀

Oooooooo can't wait - the app has been pretty much un-usable for me as a browser since forever on Pixel 5, and similar on my wife's Pixel 7.  I only use it for nightly upload/sync'ing now.

**Reply by timaschew** [reply]
@Cohinem how do you have 2 version of immich?
Are you using enter in parallel to immich? Or are you switching from ente to immich

**Reply by Cohinem** [reply]
@timaschew I basically use another app (something like [Shelter](https://f-droid.org/en/packages/net.typeblog.shelter/)) to make a clone of immich. You could also rebuild the apk with a different package name I guess.

I'm using both for now. but still prefer immich. If only its syncing were as perfect as Ente’s... hopefully in the future.

**Comment by matticas** [comment]
I know this topic maybe semi-stale, but I discovered something this morning (when I couldn't sleep at 2am).

Prior to insomnia;

- My Pixel 5; always unusably slow, can't scroll, all button clicks have 10+ second lags, etc.  Sync's perfectly fine every night - main use.  Approx 6,000 photos on phone in multiple folders sync'd.
- Wife Pixel 7; pretty damn quick after the 1.133.x update, very usable, scrolls well, bit of initial time syncing the photos and albums page (10-15seconds).  She doesn't have as much photos as me.
- Server;  1.134.0  with the new database thing.  Approx 140,000 photos, 216 albums.

During insomnia; 
- Deleted all android app data, cache and storage.  Erased all settings.
- Also removed about 4,000 old photos off the phone (already backed up in Immich of course)

Re-build steps and progress;

- Launch app, login.  Everything is extremely fast, smooth scrolling etc (i.e. it's not the server).  Albums takes some time to sync, but once open, scroll very well.
- Select the DCIM/Camera folder to Sync (about 2,000 photos).  Scrolling through the folder list to select sync's slows the App down to an unusable pace again.  All scrolling action has a 10second lag.  Once two folders were selected and eventually sync'd app becomes semi usable again.
- Found the `"[EXPERIMENTAL] Use alternate device album sync filter"` setting under Advanced.  Enabled it.  App became fast again.  Scrolling back to web user experienced, fast and usable.  App still a little slow on start up and initial sync, albums mostly.  But pressing buttons within the app are generally responsive.

Not sure if this helps any of the Dev team to improve it?  Helped my Pixel 5 a lot which was completely unusable as a GUI, only sync tool.

> > I don't know why they haven't fixed it yet.
> 
> Oh we are fixing it 😅 basically a re-write of the data structure we are using in the app so it will take a bit to get done 😀

Given the recent improvements to App GUI, manual tagging, and my ever hopes for partner face tagging / people sharing / tagging (i.e. our kids photos cross tagged from photos we both take), I have decided to support Immich more and buy a license.  

Thanks all for your work.

**Reply by matticas** [reply]
Update on 1.135 and vectorchord 0.4.2;

Unfortunately, my Android App has gone back to been very  slow and unusable.  Pretty disappointing since the progress mentioned above.  

Unsure if the source of slowness if the Android App update, or vector-chord or 1.135.x  ?

Edit;  just did a dirty test;   Web based large scrolls invoke a CPU spike but results in rapid flawless web photo scrolling.  Using my Android App on the latest version does not invoke the same CPU spike and is extremely laggy.  Using my wife's Android on an older version invokes a processor spike but is still a touch slow to respond, but eons faster then mine.

**Reply by alextran1502** [reply]
@matticas there was a but of hiccup during development (trying to juggling two databases at the same time) so the app is trying to sync from the server again. It should be operable after the sync is done, give it a couple minutes

**Reply by matticas** [reply]
> @matticas there was a but of hiccup during development (trying to juggling two databases at the same time) so the app is trying to sync from the server again. It should be operable after the sync is done, give it a couple minutes

@alextran1502   I've cleared the app cache and relaunched it, watched my server btop processors and nothing is sticking out as busy (Frigate is taking up the top 10).  No matter how long I leave it open to sync, it's still slow.  All settings menus are horribly slow etc.  

How do I locate the previous APK for Android?  I wish to rollback and run a test.

**Reply by alextran1502** [reply]
@matticas this is on the client, not on the server. So the indexing is running on your phone

**Reply by matticas** [reply]
@alextran1502   I figured as much, as my Wife's phone isn't slow.  Is it my phone per-say, or the Immich APK / app ?

**Reply by alextran1502** [reply]
Can you check if the spinning icon at the bottom is still spinning? if yes then give it sometimes

**Reply by alextran1502** [reply]
I've also created this meta issue to update you guys about this issue and how we are going to make it better

https://github.com/immich-app/immich/issues/19370

**Comment by matticas** [comment]
Hi Alex,

Yep, when the app starts, the circle spins, I wait, and once it's stops
spinning, app is still janky and all UI button presses and menus hang for
5-10 seconds per push.

**Comment by Gamrin** [comment]
Not sure if OP experiences this as fixed, but I arrived here looking for a solution to my problem.

I just turned on the beta timeline, and this seems to have improved my experience with this a lot. Even with multi-thousand albums.

**Comment by ghaisasadvait** [comment]
Improved for me as well. Backup abruptly stops sometimes, but hands down better than earlier lag

**Comment by Gamrin** [comment]
No adding to albums from timeline I the beta though. Does this mean a fix in the old timeline version shouldn't be expected? Just bear and wait till beta becomes prod?
